src/ring_theory/algebra.lean

```diff
@@ -187,19 +187,36 @@ instance of_subsemiring (S : subsemiring R) : algebra S A :=
 
 /-- Algebra over a subring. -/
 instance of_subring {R A : Type*} [comm_ring R] [ring A] [algebra R A]
+  (S : subring R) : algebra S A :=
+{ smul := λ s x, (s : R) • x,
+  commutes' := λ r x, algebra.commutes r x,
+  smul_def' := λ r x, algebra.smul_def r x,
+  .. (algebra_map R A).comp (subring.subtype S) }
+
+lemma algebra_map_of_subring {R : Type*} [comm_ring R] (S : subring R) :
+  (algebra_map S R : S →+* R) = subring.subtype S := rfl
+
+lemma coe_algebra_map_of_subring {R : Type*} [comm_ring R] (S : subring R) :
+  (algebra_map S R : S → R) = subtype.val := rfl
+
+lemma algebra_map_of_subring_apply {R : Type*} [comm_ring R] (S : subring R) (x : S) :
+  algebra_map S R x = x := rfl
+
+/-- Algebra over a set that is closed under the ring operations. -/
+instance of_is_subring {R A : Type*} [comm_ring R] [ring A] [algebra R A]
```

Comment:
Isn't there a constructions that turns a set + `is_subring` predicate into ` subring`? I think with that you can define this instance in terms of `algebra.of_subring`.

Reply:
I don't believe there is, unfortunately, only for the `is_..._hom` family. Also because the two structures have different fields, you cannot easily extend one into the other, so it would have to be manually like this:

```lean
/-- Algebra over a set that is closed under the ring operations. -/
instance of_is_subring {R A : Type*} [comm_ring R] [ring A] [algebra R A]
  (S : set R) [is_subring S] : algebra S A :=
algebra.of_subring
  ⟨S,
   is_submonoid.one_mem,
   λ a b, is_submonoid.mul_mem,
   is_add_submonoid.zero_mem,
   λ a b, is_add_submonoid.add_mem,
   λ a, is_add_subgroup.neg_mem⟩
```

I expect that we are not going to need this instance anyway as we make the subobject API consistent, so I would suggest not changing the implementation of `of_is_subring` for now.

Reply:
I think the consensus of the following topic was that `is_subring` should not be used much, but has its place in mathlib:
https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/deprecated.20is_sub*
So I do think that the following definition should be added:
```lean
def set.to_subring {α} [ring α] (S : set α) [is_subring S] : subring α :=
⟨S,
  is_submonoid.one_mem,
  λ a b, is_submonoid.mul_mem,
  is_add_submonoid.zero_mem,
  λ a b, is_add_submonoid.add_mem,
  λ a, is_add_subgroup.neg_mem⟩
```